lib/hanami/model/adapters/sql/command.rb

```diff
@@ -59,15 +55,45 @@ def update(entity)
           # @api private
           # @since 0.1.0
           def delete
-            @collection.delete
-          rescue Sequel::DatabaseError => e
-            raise Hanami::Model::Error.new(e.message)
+            _handle_database_error { @collection.delete }
           end
 
           alias_method :clear, :delete
+
+          private
+
+          # Handles any possible Adapter's Database Error
+          #
+          # @api private
+          # @since x.x.x
+          def _handle_database_error
+            yield
+          rescue Sequel::DatabaseError => e
+            raise _mapping_sequel_to_model(e)
+          end
+
+          # Maps SQL Adapter Violation's Errors into Hanami::Model Specific Errors
+          #
+          # @param adapter error [Object]
+          #
+          # @api private
+          # @since x.x.x
+          def _mapping_sequel_to_model(error)
+            case error
+            when Sequel::CheckConstraintViolation
+              Hanami::Model::CheckConstraintViolationError.new(error.message)
+            when Sequel::ForeignKeyConstraintViolation
+              Hanami::Model::ForeignKeyConstraintViolationError.new(error.message)
+            when Sequel::NotNullConstraintViolation
+              Hanami::Model::NotNullConstraintViolationError.new(error.message)
+            when Sequel::UniqueConstraintViolation
+              Hanami::Model::UniqueConstraintViolationError.new(error.message)
+            else
+              Hanami::Model::InvalidCommandError.new(error.message)
+            end
```

Comment:
On second look, what do you think about storing them into a frozen hash the query them by the key, should the pair does not match the specified key, the code fallback to the default class.

```
MAPPING = {...}.freeze

(MAPPING[error] || InvalidCommandError).new(error.message)
```